### ISSUE TITLE
3DFX Glide support?

### ISSUE THREAD

**Issue by Demorde** [stsp/dosemu2#401]
DOSEmu had some patches years ago for 3DFX/Glide on Linux through OpenGlide. Is there any possibility of getting these working again?

**Comment by stsp** [comment]
Where are they?
Did you have them working?

**Comment by stsp** [comment]
Looks like it is here:
https://github.com/voyageur/openglide/tree/master/platform/dosemu
But its not just dosemu support.
It is also a glide2x.dll that you need
to build with openwatcom and replace
the original one in the game, I suppose.
If someone has this dll, please provide
a link so that I don't need to build it.
What glide games were tested with dosemu1?

**Comment by stsp** [comment]
The readme of openglide says
`It emulates a Voodoo board`
but I don't see where does it do so.
Is anyone here in touch with s_a_white?

**Comment by stsp** [comment]
And dosbox-x seems to have the
voodoo emulation but without openglide...

**Comment by stsp** [comment]
pcem emulates voodoo too, and does not
use openglide. So I wonder what openglide
is good for. No one uses it as there seems
to be no voodoo emulator, contrary to what
its readme says...

**Comment by Demorde** [comment]
Hi!

Thanks for the response, I meant to link where I read that it had had support at one point - https://sourceforge.net/p/dosemu/feature-requests/33/

I am not sure how PCEM or DOSBox uses 3DFX, nor if you need to use OpenGlide for - I just saw mention of it and thought it'd probably easier to continue down the same road.

As for glide2x.dll you can grab it from any of the 3DFX Dosbox packages (you can use the one from nGlide - http://www.zeus-software.com/downloads/nglide or for DOSBox build http://www.si-gamer.net/gulikoza/)

As for what games worked on it, it was before my time so I am not sure what ones would work with it - I am hoping to use the Elder Scrolls Redguard (has a 3dfx demo) and Battlecruiser 3000AD (freeware).

**Comment by jschwartzenberg** [comment]
Note that there is a difference between emulating the actual 3dfx Voodoo hardware and mapping the Glide API to OpenGL using a "Glide wrapper".

Emulation of the actual hardware can be done either in software and possibly with 3D hardware acceleration as well. The first option probably won't make a big difference if the game also has a software renderer.
Using a Glide wrapper to map the calls to OpenGL generally will be good for performance obviously. Then the glide2x files need to be replaced with the ones from the wrapper instead of the original ones.

**Comment by stsp** [comment]
> As for glide2x.dll you can grab it from any of the 3DFX Dosbox packages

And that's exactly not true because
the dosemu and dosbox support are
entirely different and the sources for
glide2x.dll are there for both and are
very different.
Aside from that, it will not work with
dosemu2 by default, because, unlike
dosemu1, we use by default sdl in
software mode, not in opengl mode.
dosemu1 uses by default X backend
for which the patch was done.
So while it is possible to port this to
dosemu2 and in fact not too difficult
at all, I am quite pessimistic. People
will not compile dll, will not alter tricky
options etc.
I think if we really want something like
this, we need to steal the code from pcem.

**Comment by Demorde** [comment]
I didn't realize it would be so difficult to implement! I was just wondering since games that support 3dfx under dos usually are a lot more visually attractive if enabled.

**Comment by stsp** [comment]
Its not difficult to implement at all.
Its just that OpenGlide is nearly useless.
If someone wants to port this patch, maybe
I'll even apply it, but it is almost a waste of
time. There will be only one user of it, you. :)
Other people will not compile glide2x.dll
and will not tweak dosemu to start using opengl.

**Comment by stsp** [comment]
> Note that there is a difference between emulating the actual 3dfx Voodoo hardware and mapping the 
> Glide API to OpenGL using a "Glide wrapper".

Note that voodoo emulation is explicitly
stated in their readme. But looking more
closely, it seems this readme was created
by the person who uploaded the project
to github, not by the openglide developers.
So you can tell him about the aforementioned
differences as he looks unaware of them. :)

> The first option probably won't make a big difference if the game also has a software renderer.

It does. DOS games are not smp-aware, so
the soft rendering they do, is very limited,
without any cool effects.

> Using a Glide wrapper to map the calls to OpenGL generally will be good for performance obviously. 

Modern CPUs can handle the software voodoo
rendering without a slightest hiccup. While the
opengl problems are only becoming worse with
time. So the openglide's trade-off makes no sense.

> Then the glide2x files need to be replaced with the
> ones from the wrapper instead of the original ones.

Please compile them for us. :)
Then please adapt the patch...

**Comment by jschwartzenberg** [comment]
In that case indeed full 3dfx Voodoo emulation is the way to go. Glide wrappers were mainly made with UltraHLE and Windows games in mind (not dos) when CPUs were not that powerful. Compiling the dos Glide wrapper library is quite complex by itself. If I remember correctly one of the main reasons for the Glidos author to ask money for his work was because of all the complexity with the different compilers he needed.

**Comment by stsp** [comment]
Is pcem voodoo emulation good enough
to be worth taking a look into?

**Comment by stsp** [comment]
> UltraHLE and Windows games in mind (not dos) when CPUs were not that powerful.

Maybe for windows it indeed makes more
sense, but AFAIK opengl is a second-class
citizen there, so people use d3d. So I still
see no point.

**Comment by jschwartzenberg** [comment]
I never tried any Voodoo emulators myself. From the top of my head, MAME also has Voodoo emulation and there are also custom dosbox versions with it. (There are both dosbox versions with Glide wrapper options and versions with the actual Voodoo emulation.) There are also people looking into adding it to qemu to support games on Mac OS 9. This would probably show up on the Emaculation forums once it starts to work.

Back then OpenGL was better for nvidia cards. ATI and others indeed had D3D. Also UltraHLE ran on Wine quite early already.

**Comment by stsp** [comment]
> From the top of my head, MAME also has

non-permissive license.

> and there are also custom dosbox versions with it.

C++

> There are also people looking into adding it to qemu to support games on Mac OS 9.

That would be interesting but likely not
supporting DOS games. qemu is poor on
supporting DOS games.

**Comment by jschwartzenberg** [comment]
I think MAME switched licenses recently to fix that.

Isn't qemu's main issue for dos games the poor vga emulation?

**Comment by stsp** [comment]
> I think MAME switched licenses recently to fix that.

Good to know.
Still the C++.
But looks like mostly C.

> Isn't qemu's main issue for dos games the poor vga emulation?

Poor sb16 emulation, no DOS-specific hacks
in pic/pit/kbd code, no DOS-specific integration
with host FS, with network, etc etc.

**Comment by t9999clint** [comment]
I think the demand for Glide support is greater than you'd think. There's around 10+ games that use it, and even more if you count sourceports.
Off the top of my head...
MechWarrior 2, The TombRaider series, Shadow Warrior. Quake, Elderscrolls Redguard, These are fairly popular games too.

PCem's 3DFX emulation is getting pretty good, but it's still far from complete and is frequently getting changed. It's definately the place to start though if you plan on doing full emulation.

For passthrough however, this guy has gotten pretty far in updating qemu to get it work with both linux and windows. I think this is all c++ though so I'm not sure how much it'll help but at least they've fixed most of the openglide stuff.
https://www.vogons.org/viewtopic.php?f=24&t=60950

There are enough other very well working solutions for most of these games nowadays that this shouldn't be considered a high priority item by any means. I just figured you might find this stuff interesting.

**Comment by stsp** [comment]
Yes, the post you refer to, makes sense.
Basically they seem to claim the good speed-up
of their pass-through technique compared to the
voodoo emulation MMIO accesses.
The initial patch referred in this thread, however,
is incompatible with glide2x.dll of other emulators,
and needs a dll being built specially for dosemu. This
is unacceptable. Plus dosemu2 does not use openGL,
so we'd need the software rendering. Maybe someone
can glue OpenGlide with OpenSWR and get the sw
rendering, but I don't know how hard will this be.

So if from their architecture we can get the glide2x.dll
compatible with qemu and dosbox, and can get the
software rendering, then it would be worth considering
to add. I don't think I currently have the resources to
investigate this (and other 100+) topic.

**Comment by t9999clint** [comment]
> I don't think I currently have the resources to
> investigate this (and other 100+) topic.
I 100% understand, this is not a high priority thing at all. This would probably be more of a "well I've finished everything else I had planned, now what?", sort of thing. Basically super low priority.
Just saying that there's more than just 1 guy who would use this.... (probably just us 2 lol).

> The initial patch referred in this thread, however,
> is incompatible with glide2x.dll of other emulators,
> and needs a dll being built specially for dosemu. This
> is unacceptable. Plus dosemu2 does not use openGL,
> so we'd need the software rendering. 
I think they mentioned solving the incompatible dll issue later on in the thread, or they just compiled new ones themselves. I'm not 100% yet (I can't seem to find where it's mentioned), regardless that's not that important right now anyways.
Is there any reason why it has to be software rendering only? Would using openGL for this feature add too much to the codebase?
If it does have to be software only then using PCem's code would probably be a better plan. Also it'll probably be much closer to feature complete when there's actually enough time to get to this feature request.

**Comment by stsp** [comment]
> Would using openGL for this feature add too much to the codebase?

Its quite difficult to use sw rendering for everything
and openGL just for this feature, so that would likely
mean switching to openGL entirely. But the problem
with it is that it doesn't allow rendering from another
thread, and dosemu2 wants to do rendering from
another thread (other than the one created the window).

> If it does have to be software only then using PCem's code would probably be a better plan.

Indeed, but the main problem of it, as mentioned in
the thread you pointed, is that there are too many MMIO
accesses, so the things are slow under kvm. SW rendering
is not necessarily slow if it properly utilizes multiple CPU cores.
But emulated accesses to (non-prefetchable) MMIO regions are
definitely slow under kvm.
What do they mean by "abusive/insane LFB semantics"
is unclear (LFB is a prefetchable region, so I am puzzled).
But video cards with GPU always have a non-prefetchable
region for command stream, so I suppose this is what they talk about.

Anyway, the code from pcem has an advantage too.
Which is that the user won't need to replace any dlls in the game.
So I suppose in the ideal world the one can have both mechanisms.

**Comment by t9999clint** [comment]
I took a quick glance at the PCem voodoo code https://bitbucket.org/pcem_emulator/pcem/src/default/src/vid_voodoo.c
Now I have very little programming experiance, and even less in C (I studied Java, and C++ back in 2004), so I might be totally wrong here...

At first glance the PCem code looks fairly reasonable, it's based off c, there's proper indentation, and both variables and functions are properly named.
But the more I look at it the less good it looks, there's next to no documentation, there's no organized flow, variables and structs are created and defined at random parts throughout the entire thing, and some bits are split into other libraries and other just as complex bits are left in for no logical reason.

Added to that it's over 700kb of code, I'm starting to think PCem might be less good of an idea than I first thought... I would suggest only doing this if someone like one of the developers for PCem offered to add it for you. sa1988 and Leilei like to work on a number of different projects for this stuff so they might want to when dosemu2 gets more popular.

TL:DR, before where I only 80% agreed that this shouldn't be added right away, now I 200% agree!! Now I feel bad for resurrecting an old topic like this.

**Comment by stsp** [comment]
When porting some code that works, you
usually don't have to look inside. You just
write the wrappers to glue it to the code it
wasn't originally designed for.
Of course in this particular case, I am not
hoping for these wrappers to be small and
lean, because the video stack is usually tricky.
The main work here would be to understand
the pcem video stack and think out the adaptation
layer for dosemu2. The complexity of vid_voodoo.c
doesn't matter.

**Comment by jschwartzenberg** [comment]
https://www.youtube.com/watch?v=J0hiD3FD2Gg

**Comment by t9999clint** [comment]
Thanks for sharing that because I really loved that vid, but I'm not so sure that's a good example on why 3DFX support would be needed on DOSEMU2. There are several, better ways to play Mario 64 on Linux than in a DOS emulator. That being said it'd be really fun to run this in DOSEMU anyways.

PCem's Voodoo implementation is a lot more mature than before and is less likely to change drastically. So if you wanted to implement it in some sort of wrapper it should be a little bit easier than before. Again I'm not trying to nag, I still think this should be a fairly low priority item.

**Comment by jschwartzenberg** [comment]
Yeah of course it would just be cool :)
Quake II already runs on DOSEMU2 with software rendering.

I can imagine that the performance of games that use Glide would be much better with DOSEMU2 compared to all the other alternatives if there would be a fast Voodoo emulator, but the work seems significant.

**Comment by t9999clint** [comment]
Voodoo acceleration on DosEmu2 would easily be the fastest way to play MS-DOS glide games. Nothing else would even come close

**Comment by stsp** [comment]
AFAIK pass-through needs a custom
DOS drivers. There was an openglide
patch for dosemu1, but it was incompatible
with the one of dosbox, and so required
the different driver...
Now qemu is going to require another
one, or compatible with dosbox?

**Comment by t9999clint** [comment]
they made their own wrappers https://github.com/kjliew/qemu-3dfx/tree/master/wrapper. This is why I was saying that would be faster.
But a DOS focused hardware implementation actually make more sense if you want the best compatibility.
PCem's libraries doesn't need wrappers and can use the original code. But PCem is quite a bit different than QEmu and well, we've already discussed this at length.

**Comment by jschwartzenberg** [comment]
I didn't know anything existed beyond DOSBox/PCem. Does @kjliew' s solution on QEMU work with KVM?

**Comment by stsp** [comment]
> But a DOS focused low level emulation implementation actually make more sense if you want the best compatibility.

I don't want the best compatibility,
but what I also don't want is to force
OpenGL, with which we have problems
even w/o any 3d rendering.
But taking the pcem's impl will make
dosemu2 slowest until we avoid the
excessive PIO overhead with jit.
It should be a good start, anyway.

Does anyone know how does it
compare to the one of MAME?

**Comment by stuaxo** [comment]
Related about performance:

I thought that PCEM and MAME[1] both rendered used software to emulate the triangles in their 3DFX emulation.
There was some work in 2005 to do 3DFX rendering in MAME using hardware
https://forums.bannister.org/ubbthreads.php?ubb=showflat&Number=97436
I wonder if was ever merged ?

[1] The performance FAQ in MAME mentions setting a thread count to render triangles: https://wiki.mamedev.org/index.php/FAQ:Performance

**Comment by stsp** [comment]
> There was some work in 2005 to do 3DFX rendering in MAME using hardware

```
do all the game drawing to an off-screen framebuffer the size of the original game's,
and then use that as the source for a hardware stretchblit.
```

It basically means no HW acceleration,
and this is what we can do with SDL2
even now (except for a few troubles, so
we actually don't).
AFAIK the biggest problem with voodoo
emulation is the PIO overhead. Even if
you use HW-accelerated rendering, the
PIO overhead spoils everything.
I think pcem is trying to solve that with
some jit techniques, and we also have
jit, so at the end we could fight the PIO
overhead. But to the date, we haven't
came up with the framework to jitttify
even the VGA emulation while running
in otherwise native CPU mode.
Bart's fault-less work was a good start,
but all seems abandoned.

**Comment by stuaxo** [comment]
I probably derailed this into performance stuff, having anything that works is a good start even if slow.

Once it does maybe PIO will be more important, and good to be aware of a possible accelerated future.

**Comment by jschwartzenberg** [comment]
Not relevant when going with full hardware emulation, but someone is adding Glide support to Mesa:
https://www.phoronix.com/scan.php?page=news_item&px=2022-Mesa-Grover

Wine always had support passing Windows applications that looked for `glide2x.dll` towards `libglide2x.so.0`, so this would mainly enable those old Windows games to work.

**Comment by stsp** [comment]
I think the properly written full
hw emulation doesn't mean fully
software rasterization. I suspect
you can still use glide API to speed
up the HW emulation.
Maybe now such emulation will
materialize in some other emulator. :)
For dosemu2 this is currently out
of scope.

**Comment by uralets** [comment]
This always was the proper way to go about it (implementing the Glide-Mesa-OGL/VK bridge)

**Comment by stsp** [comment]
But for now all solutions I've seen,
involved custom driver for dos. It is
using hypercalls, not a hw access.
OTOH voodoo needs too many port
i/o for programming, so hw emulation
in this case is going to be slow anyway,
but in practice it is also using the sw
rendering.

**Comment by uralets** [comment]
Well I for one would welcome HW accelleration part of this effort.
Personally not interested as a "consumer" in all these shenanigans
emulating the correct "soft" colours (bc of some shifty bitness
conversions done by the actual cards) or authentically looking
dithering or whatever else the fanatics like to mention. All these
stuff ofc bound to slow emulation to a crawl (They'll probably do those
with shaders anyway). No, I just wish we could comfortably play some
DOS games in their Glide modes.

**Comment by stsp** [comment]
It would be good to have a hw accelled
approach, but w/o a custom dos driver.
Which is so far I've never seen.

**Comment by sjnewbury** [comment]
It occurred to me recently that it should be possible to write a DOS glide driver for dosemu2 which passed through a wrapper to the host GL stack, especially as it would be appropriate to open a new window in such a case so there would be no issue with the lack of hwaccel of the main dosemu2 vga emulation.  So I searched and quickly found this bug!

It seems this topic had plenty of opinions, but I'd like to try to see what the actual issues are according to the thread:

Cons:
OpenGL is problematic (and becoming more so with time) and OpenGLide is useless (?)
No real hardware emulation - so incompatible with software which accesses it directly (does it exist?)
VM hyper-call style interfaces are out of scope for dosemu2
No SMP support in DOS means games are uncool.
Integrating hwaccel for GLide would mean a rewrite of dosemu2 video emulation to use hwaccel

Pros:
All DOS GLide games use the GLide API, rather than access the hardware directly - by definition
Wrapper/passthrough approach gives lowest possible overhead
A separate window for GLide is appropriate and avoids architectural issues*
OpenGL support is mature, and almost certainly works on any system running dosemu2
Direct passthrough to a native GLide implementation also possible
Code already exists for dosemu1

* my suggestion

On my system at least, dosemu2 runs in vm86 mode via KVM, and uses a 32-bit command interpreter, this is already virtualisation. I really don't get the arguments about the scope of this project.  Is it simply that game support isn't the focus so something like this is just added complexity?  I suspect there is, or could be plenty of interest in dosemu2 because it provides native hardware level DOS performance, as opposed to trying to emulate a vintage PC.  To my mind, that could certainly include efficiently playing early 3D era DOS games with minimal latency which were pushing limits at the time.

**Comment by stsp** [comment]
> No real hardware emulation - so incompatible with software which accesses it directly
> (does it exist?)

But how would you force everyone
to put the dosemu2-specific glide
driver in their games?

> VM hyper-call style interfaces are out of scope for dosemu2

No, that's fundamentally not true.
But adding the dosemu2-specific
hypercalls (rather than ones that
were pre-existing) is the problem,
exactly because you'll force people
to replace the driver.

> Integrating hwaccel for GLide would mean a rewrite of dosemu2 video
> emulation to use hwaccel

Maybe not, as SDL allows you to get
an opengl context for a window, and
use it on your own.

> Code already exists for dosemu1

But is incompatible with the one of dosbox.

> I really don't get the arguments about the scope of this project.

Please quote such argument, as I
don't know what do you refer to.

>  Is it simply that game support isn't the focus so something like this is just added
> complexity?

You forget that glide is not the only
approach. You can as well steal some
voodoo impl from PCem or alike.

> To my mind, that could certainly include efficiently playing early 3D era DOS
> games with minimal latency which were pushing limits at the time.

So nobody is saying "strict no" to the
glide approach. It has obvious downsides
(forcing people to change the driver), and
there are other options to evaluate, like
stealing the HW emulation of voodoo from
somewhere.
But at the end, its all up to the implementer.
If you want to work on openglide support,
then you are more than welcome.

**Comment by stsp** [comment]
We also take a patreon donations.
Currently that's only me, but I'll be happy
to add anyone who wants to work on
dosemu2, to the donations page.
It can be you, and we'll try to find the
users who can donate to you for an
openglide support work.

**Comment by sjnewbury** [comment]
@stsp I don't *mind* working on it, I'm sure it would be fun.  I am absolutely terrible with starting projects and then moving onto something else... FWIW I have just spent a couple of weeks creating a Gentoo overlay for dosemu2 and a DJGPP cross-build environment, so I do have a decent starting point! :-)

I've spent the last couple of days trying to get current doom-legacy1 1.48 to build on DJGPP again, it's got to be easier than that!  It was the existence of the glide driver in that project which made me think about this.

**Comment by sjnewbury** [comment]
> > No real hardware emulation - so incompatible with software which accesses it directly
> > (does it exist?)
> 
> But how would you force everyone to put the dosemu2-specific glide driver in their games?
I'm under the impression, GLide drivers aren't statically linked to the game binaries, so it should just be as simple as dropping the driver into the game directory.  Maybe I'm wrong here?  I'll need to investigate this.  Last time I actually used GLide was when I got my Voodoo2 and played Mech Warrior!
> 
> > VM hyper-call style interfaces are out of scope for dosemu2
> 
> No, that's fundamentally not true. But adding the dosemu2-specific hypercalls (rather than ones that were pre-existing) is the problem, exactly because you'll force people to replace the driver.

I'm not sure how big a problem it is though.  My memories of the era were you had to engage in all kinds of hacks to get games working properly as the driver situation was in rapid flux, and would often break specific games, it's a more authentic experience!
> 
> > Integrating hwaccel for GLide would mean a rewrite of dosemu2 video
> > emulation to use hwaccel
> 
> Maybe not, as SDL allows you to get an opengl context for a window, and use it on your own.
> 

Exactly.  Create a new window with a GL context and render into that.  3DFX Voodoo1/2 hardware video out were separate from the VGA output anyway.  VGA passthrough didn't have to be connected.

> > Code already exists for dosemu1
> 
> But is incompatible with the one of dosbox.

Is the source not available?  If not, then it will obviously have to be rewritten from scratch which would be unfortunate.  
> 
> > I really don't get the arguments about the scope of this project.
> 
> Please quote such argument, as I don't know what do you refer to.
> 
I think it was probably my misinterpretation.

> > Is it simply that game support isn't the focus so something like this is just added
> > complexity?
> 
> You forget that glide is not the only approach. You can as well steal some voodoo impl from PCem or alike.

That's true, but I get the impression dosemu2 only aims to emulate what's *necessary* to present a DOS compatible environment.  QEMU/KVM/VirtIO offers a comprehensive virtualised PC/hypervisor, PCem accurate vintage PC emulation, and DOSBOX a simulation of an era accurate DOS experience.  Maybe I'm wrong about the scope of this project, but I think all the approaches have their place.

> 
> > To my mind, that could certainly include efficiently playing early 3D era DOS
> > games with minimal latency which were pushing limits at the time.
> 
> So nobody is saying "strict no" to the glide approach. It has obvious downsides (forcing people to change the driver), and there are other options to evaluate, like stealing the HW emulation of voodoo from somewhere. But at the end, its all up to the implementer. If you want to work on openglide support, then you are more than welcome.

If the decision is to go with stealing hardware emulation, wouldn't make more sense at that point to steal Voodoo3 (inc VGA) support PCem and replace the current VGA emulation.  It just seems overkill for DOS compatibility, when the overlap of DOS and advanced 3D capable GPU hardware was very limited and all existing DOS software used hw abstracted proprietary interfaces. (There were others besides 3DFX/GLide, often shipped with the sole supported game with the card!)

**Comment by stsp** [comment]
> If the decision is to go with stealing hardware emulation, wouldn't make more sense at
> that point to steal Voodoo3 (inc VGA) support PCem and replace the current VGA
> emulation.

That's a very difficult question, and
similarly as a difficult task. At least in
dosemu2 the VGA is tight to the memory
mapping subsystem, to CPU emulator
and jit, to KVM 
I'd say if you can do something w/o
touching VGA, you'd better do.

But these are not mutually exclusive things.
You can work on openglide, and even if
later the HW emulation will also be added,
no one would get sufficiently hurt.
openglide is a much easier task, so if
someone is motivated enough, I am not
going to discourage that. :)
Note that even our sdl code have changed
a lot since the last talk here, and today it
is much closer to allow the opengl accel
than it was in 2017.

**Comment by stsp** [comment]
> Is the source not available? If not, then it will obviously have to be rewritten from scratch
> which would be unfortunate.

Sources for dosbox driver are available.
But arguably starting from something
simple (which in that case would be the
dosemu1's openglide patch) is a better
way of getting something to work.

**Comment by stsp** [comment]
So whatever working solution you can
provide, is ok. Working solutions are
always good and can be changed later,
if need be.

**Comment by sjnewbury** [comment]
@stsp I actually think the current VGA emulation is good.  The only was I could envision hw accelerating it would be to map the internal state in shared GPU memory and have a compute shader perform the emulation and rendering.

I don't think my suggestion about emulating a Voodoo3 is likely a real improvement for VGA, but it would allow mixed 2D/3D as emulated hardware.  It turns out, after having a cursory search, most 3dfx compatible DOS games used a dynamically loaded GLIDE2X.OVL driver, but some had it statically linked.  So there is an argument for implementing it at the register level.  Hardware acceleration would be much easier with a wrapper though!

I'll try to find the source for the dosemu1 version and see if I can get it working.

**Comment by stsp** [comment]
It would be interesting to find out
if you can for example implement it
on register level, but instead of doing
the rendering, as PCem does, just
use the openglide wrapper for this.

**Comment by sjnewbury** [comment]
If implementing it at register level, at that point it would probably be easier to use a low level API to render, and it wouldn't be faster than doing it in software, probably slower, since the CPU would bottleneck the pipeline while having to maintain the virtual Voodoo state machine, unless you do it all in compute shader like the ParaLLEl N64 emulator: https://www.libretro.com/index.php/category/parallel-n64/

Well beyond my ability!

As I suggested above though, it might be doable for VGA.  (Not necessarily by me though! It's just an idea...)

**Comment by sjnewbury** [comment]
Found the dosemu1 bits in OpenGLide:
https://github.com/fcbarros/openglide/tree/main/platform/dosemu

**Comment by stsp** [comment]
does it work?
(with dosemu1)

**Comment by sjnewbury** [comment]
I've just reworked the plugin patch to apply to dosemu2.  Trying to build it now...

**Comment by stsp** [comment]
Oh come on!
You'd better try it on dosemu1
first, believe me. :)
Building it with dosemu2 will
likely not be the 15minutes
experience...

**Comment by sjnewbury** [comment]
@stsp Well I do have a build failure - so it is going to probably be a little more than 15mins, although I was hopeful as much of the code it touched hasn't changed that much.
```
vxd.c: In function ‘get_VXD_entry’:
vxd.c:140:26: warning: assignment to ‘unsigned int’ from ‘const unsigned int *’ makes integer from pointer without a cast [-Wint-conversion]
  140 |                     _edi = DPMI_SEL_OFF(DPMI_VXD_thunks_start) + vxd->thunk;
      |                          ^
vxd.c: In function ‘vxd_call’:
vxd.c:1860:14: warning: comparison between pointer and integer
 1860 |     if ((_eip>=1+DPMI_SEL_OFF(DPMI_VXD_thunks_start)) &&
      |              ^~
vxd.c:1861:14: warning: comparison between pointer and integer
 1861 |         (_eip<1+DPMI_SEL_OFF(DPMI_VXD_thunks_end))) {
      |              ^
vxd.c:1863:21: error: invalid operands to binary - (have ‘unsigned int’ and ‘const unsigned int *’)
 1863 |         thunks[_eip - (1+DPMI_SEL_OFF(DPMI_VXD_thunks_start))];
      |                     ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                         |
      |                         const unsigned int *
vxd.c:1865:23: warning: passing argument 1 of ‘vxd->entry’ from incompatible pointer type [-Wincompatible-pointer-types]
 1865 |           vxd->entry (scp);
      |                       ^~~
      |                       |
      |                       cpuctx_t * {aka struct pm_regs *}
vxd.c:1865:23: note: expected ‘struct sigcontext *’ but argument is of type ‘cpuctx_t *’ {aka ‘struct pm_regs *’}
vxd.c:1866:12: warning: assignment to ‘unsigned int’ from ‘const unsigned int *’ makes integer from pointer without a cast [-Wint-conversion]
 1866 |       _eip = DPMI_SEL_OFF(DPMI_VXD_thunks_end);
      |            ^

```

I'm sure it will take me longer to get it a working DOSEMU1 setup than fix the patch!

**Comment by stsp** [comment]
> Well I do have a build failure

```
diff -Naur dosemu-pristine/src/dosext/dpmi/dpmisel.h dosemu/src/dosext/dpmi/dpmisel.h
--- dosemu-pristine/src/dosext/dpmi/dpmisel.h	2009-01-30 22:55:56.000000000 +0000
+++ dosemu/src/dosext/dpmi/dpmisel.h	2009-01-30 22:59:05.000000000 +0000
@@ -48,6 +48,8 @@
 extern unsigned char	DPMI_VXD_ENABLE[];
 extern unsigned char	DPMI_VXD_APM[];
 extern unsigned char	DPMI_VXD_VTDAPI[];
+extern unsigned char    DPMI_VXD_thunks_start[];
+extern unsigned char    DPMI_VXD_thunks_end[];
```

You need to move these symbols to
`dpmisel_offsets.h`.

> I'm sure it will take me longer to get it a working DOSEMU1 setup than fix the patch!

Huge mistake.

**Comment by sjnewbury** [comment]
> > Well I do have a build failure
> 
> ```
> diff -Naur dosemu-pristine/src/dosext/dpmi/dpmisel.h dosemu/src/dosext/dpmi/dpmisel.h
> --- dosemu-pristine/src/dosext/dpmi/dpmisel.h	2009-01-30 22:55:56.000000000 +0000
> +++ dosemu/src/dosext/dpmi/dpmisel.h	2009-01-30 22:59:05.000000000 +0000
> @@ -48,6 +48,8 @@
>  extern unsigned char	DPMI_VXD_ENABLE[];
>  extern unsigned char	DPMI_VXD_APM[];
>  extern unsigned char	DPMI_VXD_VTDAPI[];
> +extern unsigned char    DPMI_VXD_thunks_start[];
> +extern unsigned char    DPMI_VXD_thunks_end[];
> ```
> 
> You need to move these symbols to `dpmisel_offsets.h`.

That's what I did.  I also changed the type to "extern const unsigned" to match the other entries, but I've obviously missed something.
> 
> > I'm sure it will take me longer to get it a working DOSEMU1 setup than fix the patch!
> 
> Huge mistake.

You say that, but I went to the trouble of creating a Gentoo Overlay with ebuilds for all of dosemu2+fdpp+djgpp+comcom32 etc to build from source because DOSEMU1 wasn't working for me! :-)

**Comment by sjnewbury** [comment]
Spotted my mistake there.  I left the type as an array, forgot to remove the "[]".

**Comment by sjnewbury** [comment]
```
vxd.c: In function ‘vxd_call’:
vxd.c:1865:23: warning: passing argument 1 of ‘vxd->entry’ from incompatible pointer type [-Wincompatible-pointer-types]
 1865 |           vxd->entry (scp);
      |                       ^~~
      |                       |
      |                       cpuctx_t * {aka struct pm_regs *}

```
This looks like it will probably break things

**Comment by stsp** [comment]
You need to adjust the patch to
use cpuctx_t - that should be a simple
change.

**Comment by stsp** [comment]
> because DOSEMU1 wasn't working for me! :-)

Wasn't working or wasn't building?
You need to try 1.4.0.8, IIRC it used to
work (sometimes).

**Comment by stsp** [comment]
How does this patch interacts with
the video output? I don't see it touching
an sdl or X back-ends, which is strange.
Does openglide create a window on its
own, or what?

**Comment by sjnewbury** [comment]
Wasn't working. There's a version in the portage tree, but it was broken, at least on an up to date system.

**Comment by stsp** [comment]
> Wasn't working.

1.4.0.8?

**Comment by sjnewbury** [comment]
> You need to adjust the patch to use cpuctx_t - that should be a simple change.

Yup.  All the modifications compile cleanly now.  Just need to add the autotoolsy stuff to the plugin.
(fails to find config.h because the plugin isn't "configured")

**Comment by sjnewbury** [comment]
dosemu-1.4.1_pre20130107-r5.ebuild

**Comment by sjnewbury** [comment]
That's the *current* portage version!

**Comment by sjnewbury** [comment]
> How does this patch interacts with the video output? I don't see it touching an sdl or X back-ends, which is strange. Does openglide create a window on its own, or what?

I didn't check yet.  From the comment in the readme I guess it creates an X11 window, or fails to... It's apparently broken. :-)

**Comment by stsp** [comment]
> I didn't check yet. From the comment in the readme I guess it creates an X11 window,
> or fails to... It's apparently broken. :-)

You mean, openglide creates?
As there seem no such code in the patch
itself.

> dosemu-1.4.1_pre20130107-r5.ebuild

I think this is some git dump.
I don't even know what is 1.4.1, but
maybe @bartoldeman does?

**Comment by sjnewbury** [comment]
Okay, looking a the plugin, first order of business is fixing this:
```

#include </home/swhite/root/include/openglide/sdk2_glide.h>
#include "glide2x.h"

```

**Comment by stsp** [comment]
> `#include </home/swhite/root/include/openglide/sdk2_glide.h>`

So maybe he used a privately patched glide.

**Comment by sjnewbury** [comment]
> > I didn't check yet. From the comment in the readme I guess it creates an X11 window,
> > or fails to... It's apparently broken. :-)
> 
> You mean, openglide creates? As there seem no such code in the patch itself.
> 
> > dosemu-1.4.1_pre20130107-r5.ebuild
> 
> I think this is some git dump. I don't even know what is 1.4.1, but maybe @bartoldeman does?

I believe it is probably the version after the last code commit, plus a good quantity of Gentoo patches:
```
dosemu-1.4.1_pre20091009-dash.patch  dosemu-1.4.1_pre20130107-fix-inline.patch  dosemu-1.4.1_pre20130107-fortify.patch       dosemu-1.4.1_pre20130107-lto.patch  dosemu-1.4.1_pre20130107-no-glibc.patch
dosemu-1.4.1_pre20130107-as.patch    dosemu-1.4.1_pre20130107-flex-2.6.3.patch  dosemu-1.4.1_pre20130107-ia16-ldflags.patch  dosemu-1.4.1_pre20130107-nm.patch
```

**Comment by sjnewbury** [comment]
> > `#include </home/swhite/root/include/openglide/sdk2_glide.h>`
> 
> So maybe he used a privately patched glide.

Possibly.  I have actually tried contacting him on sourceforge.

**Comment by stsp** [comment]
> I believe it is probably the version after the last code commit, plus a good quantity
> of Gentoo patches:

Do you have any idea why they stuck
to dosemu1 all that much? Also google
points here:
https://gpo.zugaina.org/app-emulation/dosemu2

> Possibly. I have actually tried contacting him on sourceforge.

That might be the good option.

**Comment by sjnewbury** [comment]
Amusingly the Gentoo openglide package installs "/usr/share/openglide/glide2x-dosemu.ovl"

**Comment by stsp** [comment]
Does this mean it actually worked? :)
Well, some people are very persistent
perhaps.

**Comment by sjnewbury** [comment]
> > I believe it is probably the version after the last code commit, plus a good quantity
> > of Gentoo patches:
> 
> Do you have any idea why they stuck to dosemu1 all that much? Also google points here: https://gpo.zugaina.org/app-emulation/dosemu2

Yeah, but he actual repo that points to isn't there any more.  It's a user overlay, much like my own.

I made my own.  I think part of the reason Gentoo hasn't adopted dosemu2 (yet?) is because of the various parts that a tricky to build from source, needing a djgpp cross-compiler.  Other distros can of course just ship the binaries, but Gentoo really doesn't like to do that.

> 
> > Possibly. I have actually tried contacting him on sourceforge.
> 
> That might be the good option.

**Comment by sjnewbury** [comment]
> 

Too busy chatting to actually get on with writing the autoconf files for the openglide plugin... I won't whether it works until the plugin builds!

I'll also need something to eat.. Coding on an empty stomach isn't a good idea!

**Comment by stsp** [comment]
> I won't whether it works until the plugin builds!

Don't forget to add it to `plugin_list`
file (something not needed on dosemu1).

> needing a djgpp cross-compiler

Yes, I recognize this as a problem, and
am (slowly) working on a replacement.
Since you are compiling doom with djgpp,
I wonder if you'd like to work on that
project too. :) I can already compile
comcom32 with that compiler, but it
won't run, yet. Though I never tried to
compile something as complex as doom
on it.

**Comment by stsp** [comment]
Besides, you don't need the autoconf
stuff for a plugin right away. You can as
well hard-code the libs in a Makefile for
now. Saves some time for sure.

**Comment by sjnewbury** [comment]
> Besides, you don't need the autoconf stuff for a plugin right away. You can as well hard-code the libs in a Makefile for now. Saves some time for sure.

It does need to find the openglide header and lib, so I might as well do it properly.

**Comment by sjnewbury** [comment]
> > I won't whether it works until the plugin builds!
> 
> Don't forget to add it to `plugin_list` file (something not needed on dosemu1).
> 
> > needing a djgpp cross-compiler
> 
> Yes, I recognize this as a problem, and am (slowly) working on a replacement. Since you are compiling doom with djgpp, I wonder if you'd like to work on that project too. :) I can already compile comcom32 with that compiler, but it won't run, yet. Though I never tried to compile something as complex as doom on it.

I have comcom32 working with gcc-13.2 + djgpp.  doom-legacy is a nightmare.   Two solid days of work to just get it to compile for DOS.  I'm taking a little break from that!  I did need to fix up gcc-13 to work with djgpp though, it's a bit broken out of the box.

**Comment by stsp** [comment]
https://code.launchpad.net/~jwt27/+archive/ubuntu/djgpp-toolchain/+packages

Are you aware of that repo?
Run by quite a good djgpp packager
who definitely knows what he do.

**Comment by stsp** [comment]
Although the last gcc he supports, is 12...

**Comment by sjnewbury** [comment]
I'm using his fork of djgpp. So yes, I'm aware of his work.

**Comment by sjnewbury** [comment]
> > Besides, you don't need the autoconf stuff for a plugin right away. You can as well hard-code the libs in a Makefile for now. Saves some time for sure.
> 
> It does need to find the openglide header and lib, so I might as well do it properly.

I created an adequate autoconf setup.  The plugin doesn't compile, right now.

Lots of errors relating to the change from `sigcontext` to `cpuctx_t` change.  Particularly with the DECLARE_FUNCn() macros.

**Comment by sjnewbury** [comment]
```
openglide.c:109:32: warning: ‘struct cpuctx_t’ declared inside parameter list will not be visible outside of this definition or declaration
  109 | static void THUNK(func)(struct cpuctx_t *scp) { \
      |                                ^~~~~~~~
thunks.i:1:1: note: in expansion of macro ‘DECLARE_THUNK2’
    1 | DECLARE_THUNK2(GU3DFGETINFO, FxBool, const char*, filename, Gu3dfInfo*, file_info)
      | ^~~~~~~~~~~~~~
In file included from openglide.c:5:
thunks.i: In function ‘THUNK_GU3DFGETINFO’:
../../../src/include/cpu.h:525:21: error: invalid use of undefined type ‘struct cpuctx_t’
  525 | #define _ss     (scp->ss)
      |                     ^~
openglide.c:111:69: note: in expansion of macro ‘_ss’
  111 |     unsigned long THUNK_offset = (unsigned long)LINEAR2UNIX(SEL_ADR(_ss, _esp)) - _esp; \
      |                                                                     ^~~
thunks.i:1:1: note: in expansion of macro ‘DECLARE_THUNK2’
    1 | DECLARE_THUNK2(GU3DFGETINFO, FxBool, const char*, filename, Gu3dfInfo*, file_info)
      | ^~~~~~~~~~~~~~
../../../src/include/cpu.h:487:27: error: invalid use of undefined type ‘struct cpuctx_t’
  487 | #define get_esp(s)    ((s)->esp)
      |                           ^~
[...]
```
openglide.c includes cpu.h, which is where cpuctx_t is declared, so I have no idea what's happening here.  I changed all the occurrences of sigcontext to cpuctx_t, I guess it's a difference in the structs?

**Comment by stsp** [comment]
I believe you need cpuctx_t and not
struct cpuctx_t, so remove struct.

**Comment by sjnewbury** [comment]
Produces a different error now.  It did occur to me while laying in bed last night, the patch might assume a 32-bit dosemu build.

Shall I create a [WIP] PR for this?

```
thunks.i: In function ‘THUNK_GU3DFGETINFO’:
openglide.c:113:38: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  113 |     {   unsigned long *THUNK_stack = (unsigned long *)_esp + 2; \
      |                                      ^
thunks.i:1:1: note: in expansion of macro ‘DECLARE_THUNK2’
    1 | DECLARE_THUNK2(GU3DFGETINFO, FxBool, const char*, filename, Gu3dfInfo*, file_info)
      | ^~~~~~~~~~~~~~
openglide.c:115:14: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  115 |         v1 = (t1) DWORD(THUNK_stack[0]); \
      |              ^
thunks.i:1:1: note: in expansion of macro ‘DECLARE_THUNK2’
    1 | DECLARE_THUNK2(GU3DFGETINFO, FxBool, const char*, filename, Gu3dfInfo*, file_info)
      | ^~~~~~~~~~~~~~
openglide.c:116:14: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  116 |         v2 = (t2) DWORD(THUNK_stack[1]); \
    
[...]
```
The actual OpenGLide API does appear to be properly 32-bit, even with 64-bit builds it uses 32-bit types rather than "int" so there shouldn't be any issue with thunking from 32-bit DOS.

**Comment by stsp** [comment]
Try replacing `unsigned long` with
`unsigned int`.

**Comment by sjnewbury** [comment]
I think the problem might actually be I haven't correctly, or completely converted from sigcontext to cpuctx_t.

I thought I'd fixed this yesterday, but apparently not:
```
vxd.h:14:36: warning: ‘struct cpuctx_t’ declared inside parameter list will not be visible outside of this definition or declaration
   14 |   void             (*entry)(struct cpuctx_t *scp);
      |                                    ^~~~~~~~
vxd.c: In function ‘vxd_call’:
vxd.c:1866:23: warning: passing argument 1 of ‘vxd->entry’ from incompatible pointer type [-Wincompatible-pointer-types]
 1866 |           vxd->entry (scp);
      |                       ^~~
      |                       |
      |                       cpuctx_t * {aka struct pm_regs *}

```
I haven't managed to get my head around how all this works yet.

**Comment by sjnewbury** [comment]
I think I realise what's going on.  I was missing sigcontext was a defined as a struct, while cpuctx_t is typedef struct. It makes sense now why you said to remove "struct" from openglide.c.  That also explains the compiler warning about vxd->entry (scp).

**Comment by stsp** [comment]
Its quite strange that you need to
double-guess my suggestions.
You spent a whole day just on not
doing what I have suggested to do.

**Comment by sjnewbury** [comment]
I did do it, I'm just trying to understand *what* I'm doing!  I definitely didn't double guess you. :+1:

**Comment by sjnewbury** [comment]
I haven't spent all day on this. (honestly!)  I did also try changing the unsigned longs to unsigned ints, as you suggested, but that didn't seem to work for getting rid of the warnings, it just changed which side of = was triggering the warning. I don't feel it's far from compiling successfully.  I'm going to have a close look at the preprocessor output and try to figure it out.

**Comment by sjnewbury** [comment]
@stsp, shouldn't long always be the same at ptr size on x86?

**Comment by stsp** [comment]
Yes, but ptrs under DOS are 32bit, so
we need int.
Anyway, the suggestion was insufficient.
Try this:
`unsigned int *THUNK_stack = SEL_ADR(_ss, _esp + 2);`

**Comment by sjnewbury** [comment]
Replacing the code on line 113 does make that one warning go away.  Unfortunately, many more remain.  Also, I've no idea what it's doing - yes I know it's generating functions to thunk between DOS and the OpenGlide library, but it's just making my brain hurt today! :-(  Still, I'm stubburn enough to keep trying.

Are these problems a result of the plugin assuming x86_32?  Or something else?

**Comment by stsp** [comment]
long->int are yes, but for the rest -
unlikely.

**Comment by sjnewbury** [comment]
Just to try to help me understand all this I've been chatting with ChatGPT :-)

It suggested this:
```
#define DECLARE_THUNK2(func,ret,t1,v1,t2,v2) \
static void THUNK(func)(cpuctx_t *scp) { \
    static ret FX_CALL (*THUNK_sdk) (t1, t2) = &SDK_##func; \
    uintptr_t THUNK_offset = (uintptr_t)LINEAR2UNIX(SEL_ADR(_ss, _esp)) - _esp; \
    t1 v1; t2 v2; \
    {   uintptr_t *THUNK_stack = (uintptr_t *)_esp + 2; \
        v1 = (t1) THUNK_stack[0]; \
        v2 = (t2) THUNK_stack[1]; \
    } \
    printf(__func__); \
    ret result = THUNK_sdk(v1, v2); \
    /* ... (additional code) */ \
}
```

I asked it to take into account 32-bit on the call side since it seemed to forget:
```
#define DECLARE_THUNK2(func,ret,t1,v1,t2,v2) \
static void THUNK(func)(cpuctx_t *scp) { \
    static ret FX_CALL (*THUNK_sdk) (t1, t2) = &SDK_##func; \
    uintptr_t THUNK_offset = (uintptr_t)LINEAR2UNIX(SEL_ADR(_ss, _esp)) - _esp; \
    t1 v1; t2 v2; \
    {   uintptr_t *THUNK_stack = (uintptr_t *)_esp + 2; \
        v1 = (t1) (uint32_t)THUNK_stack[0]; \
        v2 = (t2) (uint32_t)THUNK_stack[1]; \
    } \
    printf(__func__); \
    ret result = THUNK_sdk(v1, v2); \
    /* ... (additional code) */ \
}
```

**Comment by sjnewbury** [comment]
> #define DECLARE_THUNK2(func,ret,t1,v1,t2,v2) \
> static void THUNK(func)(cpuctx_t *scp) { \
>     static ret FX_CALL (*THUNK_sdk) (t1, t2) = &SDK_##func; \
>     uintptr_t THUNK_offset = (uintptr_t)LINEAR2UNIX(SEL_ADR(_ss, _esp)) - _esp; \
>     t1 v1; t2 v2; \
>     {   uintptr_t *THUNK_stack = (uintptr_t *)_esp + 2; \
>         v1 = (t1) (uint32_t)THUNK_stack[0]; \
>         v2 = (t2) (uint32_t)THUNK_stack[1]; \
>     } \
>     printf(__func__); \
>     ret result = THUNK_sdk(v1, v2); \
>     /* ... (additional code) */ \
> }

Of course it doesn't actually work ;-)

**Comment by stsp** [comment]
SEL_ADR() already does LINEAR2UNIX(),
so this is clearly wrong.

**Comment by sjnewbury** [comment]
To be fair to ChatGPT, the original macro did that too for the THUNK_offset.  I just wondered whether it could come up with a good solution taking into account the ptr length truncation.  It tried.  I has helped me somewhat to understand what's going on at least.

I wonder whether it's worth looking at what DOSBOX does?  Or are you confident of fixing this?  I admit, I'm stuggling a little with these macros, and how to handle the thunking between vm86 and Linux x86_64.

**Comment by stsp** [comment]
You need to do the change I
pointed (SEL_ADR()) and then
completely remove THUNK_offset
and __faddr(THUNK_stack).
That's just crazy.

**Comment by stsp** [comment]
You seem to have already
removed __faddr() - right, and
THUNK_offset is now unused, so
remove it as well.

**Comment by sjnewbury** [comment]
__faddr() is used throughout thunks.i so my deleting the #define was probably a mistake...?  It does use THUNK_offset though...

**Comment by sjnewbury** [comment]
```
thunks.i: In function ‘THUNK_GU3DFGETINFO’:
openglide.c:111:14: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  111 |         v1 = (t1) DWORD(THUNK_stack[0]); \
      |              ^
thunks.i:1:1: note: in expansion of macro ‘DECLARE_THUNK2’
    1 | DECLARE_THUNK2(GU3DFGETINFO, FxBool, const char*, filename, Gu3dfInfo*, file_info)
      | ^~~~~~~~~~~~~~
openglide.c:112:14: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  112 |         v2 = (t2) DWORD(THUNK_stack[1]); \
      |              ^
thunks.i:1:1: note: in expansion of macro ‘DECLARE_THUNK2’
    1 | DECLARE_THUNK2(GU3DFGETINFO, FxBool, const char*, filename, Gu3dfInfo*, file_info)
      | ^~~~~~~~~~~~~~
openglide.c:215:85: error: ‘THUNK_offset’ undeclared (first use in this function)
  215 | #define __faddr(addr) asm volatile ("mov %1, %%eax; add %%eax, %0": "=g"(addr): "m"(THUNK_offset), "0"(addr): "%eax")
      |                                                                                     ^~~~~~~~~~~~
thunks.i:3:5: note: in expansion of macro ‘__faddr’
    3 |     __faddr (filename);
      |     ^~~~~~~
openglide.c:215:85: note: each undeclared identifier is reported only once for each function it appears in
  215 | #define __faddr(addr) asm volatile ("mov %1, %%eax; add %%eax, %0": "=g"(addr): "m"(THUNK_offset), "0"(addr): "%eax")
      |                                                                                     ^~~~~~~~~~~~
thunks.i:3:5: note: in expansion of macro ‘__faddr’
    3 |     __faddr (filename);
      |     ^~~~~~~
thunks.i: In function ‘THUNK_GU3DFLOAD’:
openglide.c:111:14: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  111 |         v1 = (t1) DWORD(THUNK_stack[0]); \
      |              ^

```

This repeats for each function.

Presumably Simon had some idea what he was doing? Or has the stack handling code just changed such that those parts aren't needed any more?

**Comment by sjnewbury** [comment]
The DECLARE_THUNKn() macros look like this now, after following your suggestions:
```
#define DECLARE_THUNK2(func,ret,t1,v1,t2,v2) \
static void THUNK(func)(cpuctx_t *scp) { \
    static ret FX_CALL (*THUNK_sdk) (t1, t2) = &SDK_##func; \
    t1 v1; t2 v2; \
    {   unsigned int *THUNK_stack = (unsigned int *) SEL_ADR(_ss, _esp + 2); \
        v1 = (t1) DWORD(THUNK_stack[0]); \
        v2 = (t2) DWORD(THUNK_stack[1]); \
    } { printf (__func__);

```

**Comment by sjnewbury** [comment]
I'm not going to be able to figure this out without studying it much more, I suggest you @stsp or someone else try fixing my PR directly.  I need to cook my wifes dinner! :-)

**Comment by stsp** [comment]
> The DECLARE_THUNKn() macros look like this now, after following your suggestions:

That looks correct.
And you eliminated `__faddr(THUNK_stack);`
rightly, but there appear to be other usages
of __faddr(). For these usages try this:
`#define __faddr(addr) MEM_BASE32(addr)`
This may give some warnings, but should
still be ok.

**Comment by stsp** [comment]
Ah, there is an extra problem that
in thunks.i cpuctx_t is unavailable...
Well, yes, fixing the remaining __faddr()
is not straight-forward.

**Comment by stsp** [comment]
> @stsp or someone else try fixing my PR directly. I need to cook my wifes dinner! :-)

Possible, but it seems your PR
requires libglide2, whereas ubuntu
has libglide3.
So please fix this first.

**Comment by stsp** [comment]
Also you don't seem to be checking
for `sdk2_glide.h` and I have no idea
what is it.

**Comment by sjnewbury** [comment]
> > The DECLARE_THUNKn() macros look like this now, after following your suggestions:
> 
> That looks correct. And you eliminated `__faddr(THUNK_stack);` rightly, but there appear to be other usages of __faddr(). For these usages try this: `#define __faddr(addr) MEM_BASE32(addr)` This may give some warnings, but should still be ok.

All the problems in the plugin a *only* triggering warnings, except for:
```
openglide.c: In function ‘openglide_plugin_init’:
openglide.c:89:21: error: ‘THUNK_GRSPLASH’ undeclared (first use in this function); did you mean ‘SDK_GRSPLASH’?
   89 | #define THUNK(func) THUNK_##func
      |                     ^~~~~~
openglide.c:322:46: note: in expansion of macro ‘THUNK’
  322 |     thunks[CMD_GRSPLASH]                  = &THUNK(GRSPLASH);
      |                                              ^~~~~
openglide.c:89:21: note: each undeclared identifier is reported only once for each function it appears in
   89 | #define THUNK(func) THUNK_##func
      |                     ^~~~~~
openglide.c:322:46: note: in expansion of macro ‘THUNK’
  322 |     thunks[CMD_GRSPLASH]                  = &THUNK(GRSPLASH);

```
I haven't even looked at that error yet, since I'm pretty sure all the int<->ptr conversion warnings will probably result in runtime failure.

Anyway...
```
thunks.i: In function ‘THUNK_GU3DFGETINFO’:
openglide.c:111:14: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  111 |         v1 = (t1) DWORD(THUNK_stack[0]); \
      |              ^
thunks.i:1:1: note: in expansion of macro ‘DECLARE_THUNK2’
    1 | DECLARE_THUNK2(GU3DFGETINFO, FxBool, const char*, filename, Gu3dfInfo*, file_info)
      | ^~~~~~~~~~~~~~
openglide.c:112:14: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
  112 |         v2 = (t2) DWORD(THUNK_stack[1]); \
      |              ^
thunks.i:1:1: note: in expansion of macro ‘DECLARE_THUNK2’
    1 | DECLARE_THUNK2(GU3DFGETINFO, FxBool, const char*, filename, Gu3dfInfo*, file_info)
      | ^~~~~~~~~~~~~~
thunks.i:3:14: warning: passing argument 1 of ‘MEM_BASE32’ makes integer from pointer without a cast [-Wint-conversion]
    3 |     __faddr (filename);
      |              ^~~~~~~~
      |              |
      |              const char *
openglide.c:215:34: note: in definition of macro ‘__faddr’
  215 | #define __faddr(addr) MEM_BASE32(addr)
      |                                  ^~~~
In file included from ../../../src/include/cpu.h:11,
                 from openglide.c:5:
../../../src/include/memory.h:200:51: note: expected ‘dosaddr_t’ {aka ‘unsigned int’} but argument is of type ‘const char *’
  200 | static inline unsigned char *MEM_BASE32(dosaddr_t a) {
      |                                         ~~~~~~~~~~^
thunks.i:4:14: warning: passing argument 1 of ‘MEM_BASE32’ makes integer from pointer without a cast [-Wint-conversion]
    4 |     __faddr (file_info);
      |              ^~~~~~~~~
      |              |
      |              Gu3dfInfo *
openglide.c:215:34: note: in definition of macro ‘__faddr’
  215 | #define __faddr(addr) MEM_BASE32(addr)
      |                                  ^~~~
../../../src/include/memory.h:200:51: note: expected ‘dosaddr_t’ {aka ‘unsigned int’} but argument is of type ‘Gu3dfInfo *’
  200 | static inline unsigned char *MEM_BASE32(dosaddr_t a) {
      |                                         ~~~~~~~~~~^

```
And I really do need to go cook dinner! :-)

**Comment by stsp** [comment]
OK, please fix autoconf stuff, then I'll look.

**Comment by sjnewbury** [comment]
> Also you don't seem to be checking for `sdk2_glide.h` and I have no idea what is it.

Feel free to change that.  I'm not sure which header you're supposed to include glide.h or sdk2_glide.h.  I figured it was probably glide.h, but I didn't get around to changing openglide.c.  glide.h does include files from the openglide subdirectory though.

**Comment by sjnewbury** [comment]
I'll change it to match what openglide.c uses.
Done.

**Comment by stsp** [comment]
https://github.com/voyageur/openglide/blob/master/sdk2_glide.h
OK so this header has proprietary
copyrights, and is therefore not distributed.
Oops.
Feel free to set up the PPA for that, then
we'll see.

**Comment by sjnewbury** [comment]
I had no idea Ubuntu/Debian packages for OpenGLide don't exist. I just emerged it from the Portage tree without issue.

I'll take a look tomorrow at setting up an Ubuntu PPA, I do have it on a VM, if you're okay with it being private.

**Comment by stsp** [comment]
> I had no idea Ubuntu/Debian packages for OpenGLide don't exist. I just emerged
> it from the Portage tree without issue.

I think you emerged glide, not openglide.
Entirely different projects:
https://glide.sourceforge.net/
https://openglide.sourceforge.net/

**Comment by sjnewbury** [comment]
https://github.com/gentoo/gentoo/tree/master/media-libs/openglide

**Comment by stsp** [comment]
Then I am not sure why you don't
have sdk2_glide.h. Its in the openglide
repo. Maybe it was stripped from the
ebuild because of licensing problems?

**Comment by sjnewbury** [comment]
I do have sdk2_glide.h.  I don't know what gave you that idea?  I've updated the configure.ac to use sdk2_glide.h and it works fine for me.

**Comment by sjnewbury** [comment]
The confusion was probably because I also have a "/usr/include/glide.h" which is a symlink!

**Comment by sjnewbury** [comment]
I don't know which is "standard" to include, probably different software looks for each.

**Comment by stsp** [comment]
OK in this case only PPA is needed.
Seems like one is already there:
https://launchpad.net/~mrjingle/+archive/ubuntu/old-multimedia

**Comment by stsp** [comment]
But its outdated, so you'll need to set
up another one.

**Comment by jschwartzenberg** [comment]
> https://github.com/voyageur/openglide/blob/master/sdk2_glide.h OK so this header has proprietary copyrights, and is therefore not distributed. Oops. Feel free to set up the PPA for that, then we'll see.

Maybe they are using outdated files. At first Glide was proprietary but later on it was released under the GPL. See:
https://en.wikipedia.org/wiki/Glide_(API)
https://github.com/Danaozhong/3dfx-Glide-API

**Comment by stsp** [comment]
> Maybe they are using outdated files.

Or maybe they are using _different_ files,
that were never re-released.
If you can find exactly "sdk2_glide.h"
header under different license, then
that's interesting.

> released under the GPL. See:

Have you seen? :)
```
	"THIS SOFTWARE IS SUBJECT TO COPYRIGHT PROTECTION AND IS OFFERED 
	ONLY PURSUANT TO THE 3DFX GLIDE GENERAL PUBLIC LICENSE. THERE IS NO 
	RIGHT TO USE THE GLIDE TRADEMARK WITHOUT PRIOR WRITTEN PERMISSION OF 
	3DFX INTERACTIVE, INC. A COPY OF THIS LICENSE MAY BE OBTAINED FROM 
	THE DISTRIBUTOR OR BY CONTACTING 3DFX INTERACTIVE INC (info@3dfx.com). 
	THIS PROGRAM. IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
	EXPRESSED OR IMPLIED. SEE THE 3DFX GLIDE GENERAL PUBLIC LICENSE FOR A
	FULL TEXT OF THE NON-WARRANTY PROVISIONS.

	USE, DUPLICATION OR DISCLOSURE BY THE GOVERNMENT IS SUBJECT TO
	RESTRICTIONS AS SET FORTH IN SUBDIVISION (C)(1)(II) OF THE RIGHTS
	IN TECHNICAL DATA AND COMPUTER SOFTWARE CLAUSE AT DFARS 252.227-7013,
	AND/OR IN SIMILAR OR SUCCESSOR CLAUSES IN THE FAR, DOD OR NASA FAR
	SUPPLEMENT. UNPUBLISHED RIGHTS RESERVED UNDER THE COPYRIGHT LAWS OF
	THE UNITED STATES.

	COPYRIGHT 3DFX INTERACTIVE, INC. 1999, ALL RIGHTS RESERVED"
```
Calling that GPL is IMHO already quite
ridiculous, but they do exactly that.

**Comment by jschwartzenberg** [comment]
Yeah you're right ... it seems they wanted to exclude the trademark and made a mess...
In any case good to keep things separate from the main DOSEMU2 source tree.

**Comment by stsp** [comment]
Sure.
We have a plug-ins, that _may_ include
the proprietary headers during build, if
the user has such software installed.
If not - the plugins are not being built.
So while these things are separate from
dosemu2 source tree, the user has the
right to build them together.

**Comment by sjnewbury** [comment]
> Yeah you're right ... it seems they wanted to exclude the trademark and made a mess... In any case good to keep things separate from the main DOSEMU2 source tree.

Clearly looking at the original announcement from 3DFX it was released as free software under the 3DFX GLIDE source code GPL, not the GNU GPL.  I suppose 3DFX had other things on their mind as they were bought out by NVIDIA the following year.  APIs aren't copyrightable under US law anyway AFAIK, and I don't know what there is of 3DFX code in OpenGLide other than the headers?

**Comment by stsp** [comment]
> APIs aren't copyrightable under US law anyway AFAIK, and I don't know what
> there is of 3DFX code in OpenGLide other than the headers?

Headers is not the same thing as APIs.
They are copyrightable.

**Comment by sjnewbury** [comment]
@stsp yes, but OpenGLide could have been derived from the API, without utilising code from 3DFX and just distributed the original SDK headers unchanged for application developers to use (under the 3DFX license).  Still a bit iffy since the whole project couldn't be distributed under GNU GPL with those, but the code itself might be okay?

I'm a bit delayed getting the PPA setup since my old Launchpad account (from a couple of decades ago!) doesn't have GPG keys registered, which requires a working GPG mail client for confirmation and I normally just use my phone for email!  Apparently my KDE Akonadi/PIM installation is broken somehow...

**Comment by stsp** [comment]
> but the code itself might be okay?

Okay for what?
Its not in our source tree, so whether or
not its okay, doesn't matter much.

**Comment by sjnewbury** [comment]
I meant okay to be licensed under the GNU GPL.  I didn't mean anything more.

**Comment by stsp** [comment]
Its already under LGPL.
But I suspect you can't distribute it
with the proprietary headers, which
is probably why its not in ubuntu.

**Comment by dannycoin** [comment]
what about this:

.."AllegroGL is one of the few available opensource solutions for hardware accelerated 3D under DOS."

**ref:** https://web.archive.org/web/20170725172243/https://wiki.allegro.cc/index.php?title=Allegro_4
